mitiq/shadows/post_processing.py

```diff
@@ -0,0 +1,141 @@
+from typing import Tuple, List, Union, Any
+
+import numpy as np
+
+PAULI_MAP = {"X": 0, "Y": 1, "Z": 2}
+
+
+def snapshot_state(
+    b_list: List[Union[int, float]], u_list: List[str]
+) -> np.ndarray[Any, Any]:
+    """
+    Implement a single snapshot state reconstruction,
+
+    Args:
+        b_list (array): The list of classical outcomes for the snapshot.
+        u_list (array): Indices for the applied Pauli measurement.
+
+    Returns:
+        Numpy array with the reconstructed snapshot.
+    """
+    num_qubits = len(b_list)
+
+    # computational basis states, e.g.|0>=(1,0)
+    zero_state = np.array([[1, 0], [0, 0]])
+    one_state = np.array([[0, 0], [0, 1]])
+
+    # local unitary that applied to the computational basis states, e.g.
+    # $$Z-$$basis measurement,
+    # which equivalent to a random Pauli measurement, i.e. for each qubit,
+    # we randomly decide to measure the Pauli operators.
+    phase_z = np.array([[1, 0], [0, -1j]], dtype=np.complex128)
+    hadamard = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
+    identity = np.eye(2)
+
+    # act $$Ad_{U^\dagger}$$ on the computational basis states, where $$U$$ is
+    # Define ensemble which the random unitary sample from.
+    unitaries = [hadamard, hadamard @ phase_z, identity]
+
+    # reconstructing a single snapshot state by applying Eq. (S44)
+    rho_snapshot = np.array([1], dtype=np.complex128)
+    for i in range(num_qubits):
+        state = zero_state if b_list[i] == 1 else one_state
+        U = unitaries[PAULI_MAP[u_list[i]]]
```

Comment:
We can get rid of the index variable with [`zip`](https://docs.python.org/3.8/library/functions.html#zip)!
```suggestion
    for outcome, measurement in zip(b_list, u_list):
        state = zero_state if outcome == 1 else one_state
        U = unitaries[PAULI_MAP[measurement]]
```